### ISSUE TITLE
[macOS] Backspace collapses composing region on macOS

### ISSUE THREAD

**Issue by cbracken** [flutter/flutter#78061]
Repro steps:
1. In a textfield such as in https://github.com/cbracken/jp_flutter switch input mode to Japanese, then enter the text 日本語 (type: 'nihonngo').
2. Press backspace.

Expected result: Textfield contains the text にほん
Actual result: Textfield contains the text にほん日本

Taking a look at this under lldb:
1. The character 語 is deleted from the end of the string
2. The composing region is incorrectly collapsed to (0,0)
3. TextInputModel::UpdateComposingText is called with the new composing text (にほん)

As a result, the new composing text is injected into the string at (0,0) rather than replacing the text 日本語.

/cc @dkwingsmt since I'm wondering if this may be related to the new direct keypress handling.

Part of https://github.com/flutter/flutter/issues/30660

**Comment by cbracken** [comment]
Looking a bit further, it seems that the composing range is being reset to (0,0) in the framework.

Here’s what I'm seeing:

```
updateEditState: C: (0, 3)                          // send 0,3 to framework
setMarkedText: C: (0, 3) T: '日本語' MT: '日本語'     // setMarkedText from macOS, composing range still 0,3 in engine
updateEditState: C: (0, 3)                          // send 0,3 to framework
setEditingState: C: (0, 0)                          // receive 0,0 from framework ******
updateEditState: C: (0, 0)                          // send 0,0 to framework
setMarkedText: C: (0, 3) T: 'にほん日本' MT: 'にほん'  // setMarkedText from macOS, composing range is 0,3 in engine
updateEditState: C: (0, 3)                          // send 0,3 to framework
```

Line marked with `******` is where things have gone wrong.

**Comment by cbracken** [comment]
Verified that on backspace I'm seeing the framework send this message:
```
_setEditingState: {text: 日本, selectionBase: 2, selectionExtent: 2, selectionAffinity: TextAffinity.downstream, selectionIsDirectional: false, composingBase: -1, composingExtent: -1}
```

**Comment by cbracken** [comment]
Quick update is that the keypress event forwarding in `FlutterViewController` `dispatchKeyEvent:(NSEvent*)event ofType:(NSString*)type` is forwarding keypresses to the framework that should be handled by the IME. Ideally, while we're in composing mode, we should disable forwarding keypress events to the framework.

**Comment by dkwingsmt** [comment]
I don't quite get the description in the issue body, but I guess I understood that the issue is that the embedding should recognize whether a key event is for IME, and IME key events should not be sent to the framework, just like in Windows.
Question: should these events be directly cut off, just like in Windows, since they should have been handled before the app by the OS?

**Comment by cbracken** [comment]
> I don't quite get the repro steps in the issue body (does step 1 end up in composing mode or out of?)

In step one. you switch to Japanese input, then type the letters "nihonngo" one by one. As you type, the text will transform into kana characters as soon as possible so after "nihonng" you'll get "にほんg" in the field (depending a bit on your typing history, it may be something else wtih a "g" on the end. Then when you type the "o" you'll get 日本語. You're still in composing mode at this point, since you haven't pressed return. So at this point, still in composing mode, you press backspace. We expect the last character to be deleted and for the textfield to contain "にほん” at this point. Instead, it contains にほん日本.

I did insert some code to block sending IME keystrokes to the framework, *but* I haven't found a nice macOS API that tells you when you switch the IME to composing mode rather than direct input mode. You could try "cheating" by checking to see if `[FlutterTextInputPlugin hasMarkedText]` returns `YES` but that won't be true until after the first keystroke (e.g. "n") has been input. Since we send the events to the framework first and only hand them to the macOS responder chain once they've gone to and returned back from the framework, that check is too late.

I'll do a bit of looking to see if there's an API I can use for this. `NSTextInputContext` has a bunch of information that is *close* to what we want, but not quite *exactly* what we want.

**Comment by cbracken** [comment]
Re-assigning to @dkwingsmt since I'm unlikely to be able to take a look at this any time soon. In the digging I did, this seems like this is most likely an issue with resetting the composing region on the framework side. There are only a couple locations in the code where we either reset the composing region or replace the TextEditingValue with a new one.

**Comment by dkwingsmt** [comment]
I expect to be stuck on the final stretch of hardward keyboard as well as related keyboard bugs until at least the end of June.

**Comment by crossle** [comment]
Chinese input also faces the issue.

**Comment by justinmc** [comment]
I made a quick fix for a similar problem just by controlling which keys are passed to the platform by the framework: https://github.com/flutter/flutter/pull/85335

Kind of like @cbracken mentioned at the bottom of https://github.com/flutter/flutter/issues/78061#issuecomment-798794150.  Not sure if that PR is going to pass all the tests and end up being a solid solution, but these problems are definitely related.

**Comment by Heasn** [comment]
any update?

**Comment by boyan01** [comment]
After research. I found that this issue occurs because flutter receive keyborad events before the input method on macOS.

For example. when the TextField is in the Composing state and then user presses the `backspace` key, the flutter framework will consume the delete event, so the composing text on the compsing floating window cannot be deleted.

Fortunately, all actions(delete, forward) in the `TextFiled` in flutter can be overridden, so I wrote a simple workaround for this issue.

Here is the code of the workaround.  https://github.com/MixinNetwork/flutter-app/pull/535

And it seems works nice on my app.

https://user-images.githubusercontent.com/17426470/155689107-69b09fd6-e295-4165-a1e3-76af00173806.mov

------

Hope ths workaround helps the guy who is suffering from this issue.

**Comment by boyan01** [comment]
> This should be fixed by [flutter/engine#32051](https://github.com/flutter/engine/pull/32051) and should be available on the framework master branch soon.

nice work!! 

But there seems to be a small problem here. I post the details to this new issue https://github.com/flutter/flutter/issues/100272

**Comment by cbracken** [comment]
@boyan01 thanks for filing! We'll take a look!
